beacon-chain/state/getters.go

```diff
@@ -647,39 +628,256 @@ func CopyPendingAttestation(att *pbp2p.PendingAttestation) *pbp2p.PendingAttesta
+func CopyIndexedAttestation(indexedAtt *ethpb.IndexedAttestation) *ethpb.IndexedAttestation {
+	var indices []uint64
+	if indexedAtt == nil {
+		return nil
+	} else if indexedAtt.AttestingIndices != nil {
+		indices = make([]uint64, len(indexedAtt.AttestingIndices))
+		copy(indices[:], indexedAtt.AttestingIndices)
+	}
+	return &ethpb.IndexedAttestation{
+		AttestingIndices: indices[:],
+		Data:             CopyAttestationData(indexedAtt.Data),
+		Signature:        safeCopyBytes(indexedAtt.Signature),
+	}
+}
+
+// CopyAttestations copies the provided Attestation array.
+func CopyAttestations(attestations []*ethpb.Attestation) []*ethpb.Attestation {
+	if attestations == nil {
+		return nil
+	}
+	newAttestations := make([]*ethpb.Attestation, len(attestations))
+	for i, att := range attestations {
+		newAttestations[i] = CopyAttestation(att)
+	}
+	return newAttestations[:]
+}
+
+// CopyDeposits copies the provided deposit array.
+func CopyDeposits(deposits []*ethpb.Deposit) []*ethpb.Deposit {
+	if deposits == nil {
+		return nil
+	}
+	newDeposits := make([]*ethpb.Deposit, len(deposits))
+	for i, dep := range deposits {
+		newDeposits[i] = CopyDeposit(dep)
+	}
+	return newDeposits[:]
+}
+
+// CopyDeposit copies the provided deposit.
+func CopyDeposit(deposit *ethpb.Deposit) *ethpb.Deposit {
+	if deposit == nil {
+		return nil
+	}
+	return &ethpb.Deposit{
+		Proof: copy2dBytes(deposit.Proof),
+		Data:  CopyDepositData(deposit.Data),
+	}
+}
+
+// CopyDepositData copies the provided deposit data.
+func CopyDepositData(depData *ethpb.Deposit_Data) *ethpb.Deposit_Data {
+	if depData == nil {
+		return nil
+	}
+	return &ethpb.Deposit_Data{
+		PublicKey:             safeCopyBytes(depData.PublicKey),
+		WithdrawalCredentials: safeCopyBytes(depData.WithdrawalCredentials),
+		Amount:                depData.Amount,
+		Signature:             safeCopyBytes(depData.Signature),
+	}
+}
+
+// CopySignedVoluntaryExits copies the provided SignedVoluntaryExits array.
+func CopySignedVoluntaryExits(exits []*ethpb.SignedVoluntaryExit) []*ethpb.SignedVoluntaryExit {
+	if exits == nil {
+		return nil
+	}
+	newExits := make([]*ethpb.SignedVoluntaryExit, len(exits))
+	for i, exit := range exits {
+		newExits[i] = CopySignedVoluntaryExit(exit)
+	}
+	return newExits[:]
+}
+
+// CopySignedVoluntaryExit copies the provided SignedVoluntaryExit.
+func CopySignedVoluntaryExit(exit *ethpb.SignedVoluntaryExit) *ethpb.SignedVoluntaryExit {
+	if exit == nil {
+		return nil
+	}
+
+	return &ethpb.SignedVoluntaryExit{
+		Exit: &ethpb.VoluntaryExit{
+			Epoch:          exit.Exit.Epoch,
+			ValidatorIndex: exit.Exit.ValidatorIndex,
+		},
+		Signature: safeCopyBytes(exit.Signature),
+	}
+}
+
+func safeCopyBytes(cp []byte) []byte {
+	if cp != nil {
+		copied := make([]byte, len(cp))
+		copy(copied, cp)
+		return copied
+	}
+	return nil
+}
+
+func copy2dBytes(ary [][]byte) [][]byte {
+	copied := make([][]byte, len(ary))
+
+	for i := 0; i < len(ary); i++ {
```

Comment:
```suggestion
		for i,a := range ary {
		copied[i] = safeCopyBytes(a)
	}

```
should be more straightforward